PROD-398 Add device ID for Identity management

neel-ds opened the pull request:
### Description
This PR implements server-side Identity Management to track events or UTM for pre-login and post-login users. It will allow us to track both kinds of users and connect the identities. Helpful in lead conversion and other metrics. `$device_id` will be a common attribute between pre-login and post-login.
### What are the steps to test that this code is working?
Open various pages with UTM parameters and interact with them. Then, log in and interact with the reveal feature. After that, check the Mixpanel dashboard to ensure those events are being tracked correctly.
### Screenshots or recordings for UI changes
Pre-login
![Screenshot 2024-10-22 at 4 48 37 PM](https://github.com/user-attachments/assets/8277a036-0f0e-4a8c-a7fb-1f04e1f95b52)
Post-login
![Screenshot 2024-10-22 at 4 48 27 PM](https://github.com/user-attachments/assets/b8e1c06f-db1e-4850-83e0-568c0d74296d)

marvinmarnold commented on app/api/mixpanel/route.ts:
```diff
@@ -32,24 +77,51 @@ export async function POST(request: NextRequest) {
-        utm_content: properties.$utm_content,
-      };
+      // Post-login user tracking
+      const { initialUtm, lastUtm } = await handleUtmParams(currentUser.id, properties);
```
User UTM should be used if present, otherwise use device UTM.

neel-ds replied:
Done

marvinmarnold reviewed (changes requested):
Can you add more test cases for all the scenarios we talked about.
- normal: logged out user visits, then confirms
- slow conversion: guest visits from Twitter, then later from Instagram and signs up
- logged in users: codify in a test what the behavior should be if user is logged in first time UTM is tracked, then signs out from the same device

Make clear assertions about the state of first and last utm for each case.

Let's get this into the next release.

neel-ds commented:
> Can you add more test cases for all the scenarios we talked about.
> 
> * normal: logged out user visits, then confirms
> * slow conversion: guest visits from Twitter, then later from Instagram and signs up
> * logged in users: codify in a test what the behavior should be if user is logged in first time UTM is tracked, then signs out from the same device
> 
> Make clear assertions about the state of first and last utm for each case.
> 
> Let's get this into the next release.

I added three more test cases specific to UTM tracking which asserts edge cases as mentioned above. Fixed some mistakes in route as well. Refactored codebase to meet standard practices as suggested.

yash431garg commented on app/api/mixpanel/route.ts:
```diff
@@ -23,6 +25,57 @@ function getIPAddress(headersList: ReadonlyHeaders) {
+  );
+
+  // Update initial and last UTM data
+  if (hasNewUtmParams) {
```
Here you can directly check from (utmParamsWithoutUndefined) that if some value is defined and you don't need to do this extra step (hasNewUtmParams).

neel-ds replied:
Good catch!

yash431garg commented on app/api/mixpanel/route.ts:
```diff
@@ -32,77 +85,66 @@ export async function POST(request: NextRequest) {
+    cookies().set('device_id', deviceId, { 
+      httpOnly: true, 
+      secure: process.env.NODE_ENV === 'production',
+      maxAge: 60 * 60 * 24 * 365 // 1 YEAR
```
Not sure about the reason behind it. But Isn't 1year to long?

neel-ds replied:
Default behavior: If you don't set a maxAge or expires attribute, the cookie becomes a "session cookie". This means it will be deleted when the browser session ends (i.e., when the user closes their browser).

So ideally to have trace for longer period, it's good to have a year even if user revisits in months or so

yash431garg replied:
Thanks for the clarification!

neel-ds commented on app/api/mixpanel/route.ts:
```diff
@@ -23,6 +25,57 @@ function getIPAddress(headersList: ReadonlyHeaders) {
+    (value) => value !== undefined,
+  );
+
+  const utmParamsWithoutUndefined = Object.fromEntries(
```
Thanks for the suggestion

marvinmarnold commented on __tests__/api/mixpanel/route.test.ts:
```diff
@@ -0,0 +1,291 @@
+      ],
+    };
+
+    await kv.set("utm:550e8400-e29b-41d4-a716-446655440000", {});
```
What is the purpose of these lines?

neel-ds replied:
The purpose of these lines is to clear the values for deviceId and userId before each test. If previous values remain, they are mistakenly treated as initial values in the subsequent tests.

marvinmarnold commented on __tests__/api/mixpanel/route.test.ts:
```diff
@@ -0,0 +1,291 @@
+      last_utm: { utm_source: "twitter", utm_medium: "social" },
+    });
+  });
+});
```
Could you add a test case where:
1. user is logged out to start
2. then logs in (assert UTM carries over)
3. user logs in from a different device (assert UTM carries over)

neel-ds replied:
Added!
